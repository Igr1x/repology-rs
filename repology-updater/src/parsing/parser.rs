// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use crate::package::ParsedPackage;
use crate::parsing::error::PackageParsingError;
use crate::parsing::package_maker::PackageMaker;

pub trait PackageSink {
    fn push(&mut self, package_maker: PackageMaker) -> Result<(), PackageParsingError>;
}

pub trait PackageParser {
    fn parse(&self, path: &Path, sink: &mut dyn PackageSink) -> anyhow::Result<()>;
}

#[derive(Default)]
pub struct PackageAccumulator {
    pub packages: Vec<ParsedPackage>,
}

impl PackageSink for PackageAccumulator {
    fn push(&mut self, package_maker: PackageMaker) -> Result<(), PackageParsingError> {
        self.packages.push(package_maker.finalize()?);
        Ok(())
    }
}
