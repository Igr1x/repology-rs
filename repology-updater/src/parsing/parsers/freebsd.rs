// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use anyhow::{Context, anyhow, bail};
use std::io::BufRead as _;
use std::path::Path;

use crate::package::Package;
use crate::parsing::error::PackageParsingError;
use crate::parsing::package_maker::PackageMaker;
use crate::parsing::parser::{PackageParser, PackageSink};

pub struct FreeBsdParser {}

impl FreeBsdParser {
    fn parse_line(line: &str, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        let fields: Vec<_> = line.trim().split('|').collect();

        const EXPECTED_FIELDS_COUNT: usize = 13;

        if fields.len() != EXPECTED_FIELDS_COUNT {
            bail!(
                "expected {} fields, got {}",
                EXPECTED_FIELDS_COUNT,
                fields.len()
            );
        }

        let mut pkg = PackageMaker::default();

        let (name, version) = fields[0]
            .rsplit_once('-')
            .ok_or_else(|| anyhow!("expected <package name>-<version> in the first field"))?;

        pkg.projectname_seed = Some(name.to_owned());
        pkg.version = Some(version.to_owned());

        Ok(sink.push(pkg)?)
    }
}

impl PackageParser for FreeBsdParser {
    fn parse(&self, path: &Path, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        let f = std::fs::File::open_buffered(path)?;

        for (nline, line) in f.lines().enumerate() {
            Self::parse_line(&line?, sink).with_context(|| format!("on line {}", nline + 1))?;
        }

        Ok(())
    }
}

#[cfg(test)]
#[coverage(off)]
mod tests {
    use super::*;

    use tempfile::tempdir;

    use std::io::Write as _;

    use crate::parsing::parser::PackageAccumulator;

    #[test]
    fn test_parse_ok() {
        let dir = tempdir().unwrap();
        let index_path = dir.path().join("INDEX");
        writeln!(
            std::fs::File::create(&index_path).unwrap(),
            "trojka-1.0_2|/usr/ports/games/trojka|/usr/local|Game of skill|/usr/ports/games/trojka/pkg-descr|e@ik.nu|games||||||"
        ).unwrap();

        let mut packages = PackageAccumulator::default();
        FreeBsdParser {}.parse(&index_path, &mut packages).unwrap();

        assert_eq!(
            packages.packages[0],
            Package {
                projectname_seed: "trojka".to_owned(),
                version: "1.0_2".to_owned()
            }
        );
    }

    #[test]
    fn test_parse_fail() {
        let cases = vec![
            // missing field
            "trojka-1.0_2|/usr/ports/games/trojka|/usr/local|Game of skill|/usr/ports/games/trojka/pkg-descr|e@ik.nu|games|||||",
            // extra field
            "trojka-1.0_2|/usr/ports/games/trojka|/usr/local|Game of skill|/usr/ports/games/trojka/pkg-descr|e@ik.nu|games|||||||",
            // bad package name format
            "trojka1.0_2|/usr/ports/games/trojka|/usr/local|Game of skill|/usr/ports/games/trojka/pkg-descr|e@ik.nu|games||||||",
            // no version
            "trojka-|/usr/ports/games/trojka|/usr/local|Game of skill|/usr/ports/games/trojka/pkg-descr|e@ik.nu|games||||||",
            // no name
            "-1.0_2|/usr/ports/games/trojka|/usr/local|Game of skill|/usr/ports/games/trojka/pkg-descr|e@ik.nu|games||||||",
        ];

        for case in cases {
            let dir = tempdir().unwrap();
            let index_path = dir.path().join("INDEX");
            writeln!(std::fs::File::create(&index_path).unwrap(), "{case}",).unwrap();

            let mut _packages = PackageAccumulator::default();
            let res = FreeBsdParser {}.parse(&index_path, &mut _packages);
            assert!(res.is_err());
            eprintln!("{:#?}", res);
        }
    }
}
