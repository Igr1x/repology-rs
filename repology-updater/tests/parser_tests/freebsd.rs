// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use repology_updater::parsing::parser::PackageAccumulator;
use repology_updater::parsing::parser::PackageParser as _;
use repology_updater::parsing::parsers::freebsd::FreeBsdParser;

#[test]
#[ignore]
fn test_freebsd() {
    let mut packages = PackageAccumulator::default();
    FreeBsdParser {}
        .parse(Path::new("tests/parser_tests/freebsd/ok"), &mut packages)
        .unwrap();
    insta::assert_debug_snapshot!(packages.packages);
}

#[test]
fn test_error_missing_field() {
    let mut packages = PackageAccumulator::default();
    insta::assert_debug_snapshot!(FreeBsdParser {}.parse(
        Path::new("tests/parser_tests/freebsd/error1"),
        &mut packages,
    ))
}

#[test]
fn test_error_extra_field() {
    let mut packages = PackageAccumulator::default();
    insta::assert_debug_snapshot!(FreeBsdParser {}.parse(
        Path::new("tests/parser_tests/freebsd/error2"),
        &mut packages,
    ))
}

#[test]
fn test_error_missing_version() {
    let mut packages = PackageAccumulator::default();
    insta::assert_debug_snapshot!(FreeBsdParser {}.parse(
        Path::new("tests/parser_tests/freebsd/error3"),
        &mut packages,
    ))
}

#[test]
fn test_error_missing_name() {
    let mut packages = PackageAccumulator::default();
    insta::assert_debug_snapshot!(FreeBsdParser {}.parse(
        Path::new("tests/parser_tests/freebsd/error4"),
        &mut packages,
    ))
}

#[test]
fn test_error_bad_package_format() {
    let mut packages = PackageAccumulator::default();
    insta::assert_debug_snapshot!(FreeBsdParser {}.parse(
        Path::new("tests/parser_tests/freebsd/error5"),
        &mut packages,
    ))
}

#[test]
fn test_error_bad_package_path() {
    let mut packages = PackageAccumulator::default();
    insta::assert_debug_snapshot!(FreeBsdParser {}.parse(
        Path::new("tests/parser_tests/freebsd/error6"),
        &mut packages,
    ))
}
