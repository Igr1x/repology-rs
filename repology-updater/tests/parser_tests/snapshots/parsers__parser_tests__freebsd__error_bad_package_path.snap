---
source: repology-updater/tests/parser_tests/freebsd.rs
expression: "FreeBsdParser\n{}.parse(Path::new(\"tests/parser_tests/freebsd/error6\"), &mut packages,)"
---
Err(
    Error {
        context: "on line 1",
        source: "unexpectedly short port path (expected at least category/name)",
    },
)
